File: proj/FkAndroid/fkpreauto/src/main/java/com/filmkilns/preauto/core/FkNativeObject.java
package com.filmkilns.preauto.core;

abstract public class FkNativeObject {
    public interface Task {
        int run(long handle);
    }

    private long handle = 0L;

    public boolean isNull() {
        return handle == 0L;
    }

    public abstract long onNativeCreate();

    public abstract boolean onNativeDestroy(long handle);

    public int create() {
        long _handle = onNativeCreate();
        synchronized (this) {
            if (_handle != 0L) {
                handle = _handle;
                return 0;
            }
        }
        return -1;
    }

    public int destroy() {
        synchronized (this) {
            if (isNull()) {
                return -1;
            }
            long _handle = handle;
            handle = 0L;
            boolean state = onNativeDestroy(_handle);
            if (state) {
                return 0;
            }
        }
        return -1;
    }

    public int with(Task task) {
        synchronized (this) {
            if (isNull()) {
                return -1;
            }
            return task.run(handle);
        }
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        destroy();
    }
}

File: proj/FkAndroid/fkandroid/src/main/java/com/alimin/fk/core/FkNativeObject.kt
package com.alimin.fk.core

import com.alimin.fk.entity.FkResult

abstract class FkNativeObject {
    private var handle = 0L
    fun isNull(): Boolean = handle == 0L
    abstract fun onNativeCreate(): Long
    abstract fun onNativeDestroy(handle: Long): Boolean
    open fun create(): FkResult {
        val _handle = onNativeCreate()
        synchronized(this) {
            if (_handle != 0L) {
                handle = _handle
                return FkResult.OK
            }
        }
        return FkResult.INVALID_DATA
    }

    open fun destroy(): FkResult {
        synchronized(this) {
            if (isNull()) {
                return FkResult.INVALID_STATE
            }
            val _handle = handle
            handle = 0L
            val state = onNativeDestroy(_handle)
            if (state) {
                return FkResult.OK
            }
        }
        return FkResult.FAIL
    }

    fun with(task: (handle: Long) -> FkResult): FkResult {
        synchronized(this) {
            if (isNull()) {
                return FkResult.INVALID_DATA
            }
            return task(handle)
        }
    }

    protected fun finalize() {
        destroy()
    }
}